spec/src/main/asciidoc/asynchronous.asciidoc

```diff
@@ -52,7 +52,15 @@ public CompletionStage<Connection> serviceA() {
 The above code-snippet means that the `Asynchronous` policy is applied to the `serviceA` method, which means that a call to `serviceA` will return a `CompletionStage` immediately and that execution of the method body will be done on a different thread.
 
 The `@Asynchronous` annotation can be used together with `@Timeout`, `@Fallback`, `@Bulkhead`, `@CircuitBreaker` and `@Retry`.
-In this case, the method invocation and any fault tolerance processing will occur in a different thread. The returned `Future` or `CompletionStage` will be completed with the final result once all other Fault Tolerance processing has been completed.
+In this case, the method invocation and any fault tolerance processing will occur in a different thread. The returned `Future` or `CompletionStage` will be completed with the final result once all other Fault Tolerance processing has been completed. However, the two different return type have some differences, listed below.
+
+==== Return a Future
+If a `Future` completes exceptionally, the exception may not trigger other Fault Tolerance capabilities, even if specified. For instance, if a method returns a `Future` with `Asynchronous` and `Retry` specified, the `Future` that completes exceptionally will not trigger the method to be retried. 
+
+==== Return a CompletionStage 
+When a method returns a `CompletionStage`, if the `CompletionStage` returns exceptionally, the exception will trigger other Fault Tolerance capabilities, e.g. `Retry` if specified. Therefore, the usage of `CompletionStage` is more preferred.
+
+Therefore, it is highly encouraged to use `CompletionStage` instead of `Future` as a method with the return of `Future` behaves similar to synchronous calls.
```

Comment:
I don't think this section is very clear. It needs to clearly show the differences between the behaviour when using `CompletionStage` and when using `Future`.

I've put together some alternate text, what do you think?

=== Interaction with other fault tolerance annotations

`@Asynchronous` can be used in conjunction with any of the other Fault Tolerance annotations, but the behavior is slightly different depending on whether the method returns a `Future` or a `CompletionStage`.

==== Interactions when returning a Future

If the method returns a `Future`, then the other fault tolerance policies are applied only around the method call itself. This means:
* If the method throws an exception then that exception may trigger other fault tolerance logic.
* If the method returns a `Future` then the method call will be considered to have succeeded. This is the case even if the `Future` that was returned is incomplete, was cancelled or would throw an `ExecutionException` if its `get()` method was called.

==== Interactions when returning a CompletionStage 

If the method returns a `CompletionStage`, then the other fault tolerance policies are applied only around both the method call, and the eventual result of the `CompletionStage`. This means:
* If the method throws an exception then that exception may trigger other fault tolerance logic.
* If the method returns a `CompletionStage` which has completed exceptionally, then that exception may trigger other fault tolerance logic.
* If the method returns a `CompletionStage` which is incomplete, but then later completes exceptionally, then that exception may trigger other fault tolerance logic at the time when the `CompletionStage` completes excpetionally.
* The method call is only considered to have completed successfully if the method returns a `CompletionStage` and that `CompletionStage` has completed successfully.

This behaviour makes it easy to apply fault tolerance logic around a `CompletionStage` which was returned from another component, for example applying `@Retry` and `@Timeout` to the result of a  JAX-RS client call.

Reply:
On re-reading my suggestion, I don't like they way I've used the word "may".

I'm not sure how best to express the idea that any given call either succeeds, or fails with an exception and that other fault tolerance policies, depending on their configuration, may take some action depending on the exception with which the method failed.

Reply:
@Emily-Jiang I think we should use the text proposed by @Azquelt. It's has more detail.
I'm not really sure about:

"If the method returns a CompletionStage which is incomplete, but then later completes exceptionally, then that exception may trigger other fault tolerance logic at the time when the CompletionStage completes excpetionally."

It's not clear which "other fault tolerance logic" we are talking. We should mention, "the other fault tolerance logic in the method with the original `@Asynchronous` annotation", or something around those lines.